src/lib.rs

```diff
@@ -62,41 +65,88 @@ pub fn run(args: Args) {
-                        args_filter_arg,
-                    );
-
-                    SearcherBuilder::new()
-                        .multi_line(true)
-                        .build()
-                        .search_path(&matcher, path, printer.sink_with_path(&matcher, path))
-                        .unwrap();
-                }
+    get_project_file_walker(&*supported_language)
+        .into_parallel_iterator()
+        .for_each({
+            |project_file_dir_entry| {
+                let mut printer = grep::printer::Standard::new_no_color(io::stdout());
+                let path = project_file_dir_entry.path();
+
+                let matcher = TreeSitterMatcher::new(
+                    &query,
+                    capture_index,
+                    language,
+                    args.filter.clone(),
+                    args.filter_arg.clone(),
+                );
+
+                SearcherBuilder::new()
+                    .multi_line(true)
+                    .build()
+                    .search_path(&matcher, path, printer.sink_with_path(&matcher, path))
+                    .unwrap();
+            }
+        });
+}
+
+trait IntoParallelIterator {
+    fn into_parallel_iterator(self) -> IterBridge<WalkParallelIterator>;
+}
+
+impl IntoParallelIterator for WalkParallel {
+    fn into_parallel_iterator(self) -> IterBridge<WalkParallelIterator> {
+        WalkParallelIterator::new(self).par_bridge()
+    }
+}
+
+struct WalkParallelIterator {
+    receiver_iterator: <Receiver<DirEntry> as IntoIterator>::IntoIter,
+    _handle: JoinHandle<()>,
+}
+
+impl WalkParallelIterator {
+    pub fn new(walk_parallel: WalkParallel) -> Self {
+        let (sender, receiver) = mpsc::channel::<DirEntry>();
+        let handle = thread::spawn(move || {
+            walk_parallel.run(move || {
+                Box::new({
+                    let sender = sender.clone();
+                    move |entry| {
+                        let entry = match entry {
+                            Err(_) => return WalkState::Continue,
+                            Ok(entry) => entry,
+                        };
+                        if !entry.metadata().unwrap().is_file() {
+                            return WalkState::Continue;
+                        }
+                        sender.send(entry).unwrap();
+                        WalkState::Continue
+                    }
+                })
             });
+        });
+        Self {
+            receiver_iterator: receiver.into_iter(),
+            _handle: handle,
         }
-        thread_pool.join();
-    });
-    walk_project_files(&*supported_language, sender);
-    handle.join().unwrap();
+    }
+}
+
+impl Iterator for WalkParallelIterator {
+    type Item = DirEntry;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.receiver_iterator.next() {
+            Some(item) => Some(item),
+            None => {
+                // self.handle.join().unwrap();
```

Comment:
Not sure what my mental model is/should be here - we've definitely "exhausted" this encapsulated thread because the owned receiver iterator is returning `None` so therefore all senders have been dropped (ie the `walk_parallel.run()` callback has been dropped) so therefore `walk_parallel` must basically be done and most likely has already returned (ie the thread has already finished)? So then this `.join()` shouldn't block for long if at all? But by the same token is there any reason to explicitly `.join()` at that point (my understanding is that if the main thread finished `main()` before that thread had actually finished then it would just silently be killed)?

Reply:
Yeah, your reasoning makes sense to me there. If you're confident that the thread will be cleanup up (either by finishing, or the parent process terminating), I don't really see a reason to join it.

Reply:
Ok per other comment then reduced this down to just a "pure delegation" to `self.receiver_iterator.next()`